feat!: Support both YAML and HCL config

disq opened the pull request.

disq committed:
feat!: Support YAML config with HCL

disq commented on provider/provider.go:
```diff
@@ -113,31 +140,36 @@ func (p *Provider) ConfigureProvider(_ context.Context, request *cqproto.Configu
+				Diagnostics: diags,
+			}, nil
+		}
 	}
```
TODO do the same for ConfigYAML

disq committed:
cleanly support both formats

disq committed:
Use correct yaml pkg, make yaml default

roneli commented on provider/provider.go:
```diff
@@ -113,23 +140,32 @@ func (p *Provider) ConfigureProvider(_ context.Context, request *cqproto.Configu
 	}
+
 	// if we received an empty config we notify in log and only use defaults.
 	if len(request.Config) == 0 {
```
Looking good, some thoughts, tell me what you think about them:

why not let in the request the core passes the format of configuration? prior cores by default its set to HCL and used. new cores can decide based on configuration passed. 

Should the "provider" care / pass what format is supports? it just needs to know how to decode the block and pass as interface to main Configure function. I think also in the GetProviderConfig request we should do the same we can request yaml/hcl format if provider doesn't support yaml format we default back to hcl.

roneli reviewed (commented):
Looking good, didn't approve as this is a draft, added some of my thoughts, tell me what you think :)

disq committed:
ExampleFormat => Format

disq committed:
updates

roneli reviewed (commented):
Looks good, obviously not finished, but didn't have much comments one nit. yaml.Node is a great  solution.

disq committed:
do yaml

disq commented on provider/provider.go:
```diff
@@ -71,19 +84,48 @@ func (p *Provider) GetProviderSchema(_ context.Context, _ *cqproto.GetProviderSc
 			%s
 			// list of resources to fetch
 			resources = %s
-		}`, p.Name, p.Config().Example(), helpers.FormatSlice(funk.Keys(p.ResourceMap).([]string)))
```
IMO that would affect readability adversely (if const it has to be up top, we have multiple params, etc)

disq committed:
updates

yevgenypats commented on provider/provider.go:
```diff
@@ -71,19 +78,70 @@ func (p *Provider) GetProviderSchema(_ context.Context, _ *cqproto.GetProviderSc
+		}, nil
+	case cqproto.ConfigYAML:
+		var data ProviderConfiguration
+		if err := yaml.Unmarshal([]byte(providerConfig.Example()), &data.Inline); err != nil {
```
why the unmarshal/marshal is necessary here? why not just send the []byte/string as is?

disq replied:
because we add resource list to the bottom in line 105..130

yevgenypats reviewed (approved):
Looks good. just had one question/clarification.

disq committed:
Make it work

disq committed:
remove one level

yevgenypats commented on provider/provider.go:
```diff
@@ -71,19 +78,59 @@ func (p *Provider) GetProviderSchema(_ context.Context, _ *cqproto.GetProviderSc
+		}, nil
+	case cqproto.ConfigYAML:
+		var data ProviderConfiguration
+		if err := yaml.Unmarshal([]byte(providerConfig.Example()), &data); err != nil {
```
Why this unmarshal/marshal is necessary ?

roneli reviewed (approved):
LGTM, no further comments :)

disq committed:
Always create ProviderConfiguration.Inline

disq committed:
GetProviderConfig: generate YAML with string operations(!)

This way it's possible to maintain comments coming in from the provider.Example()

disq committed:
Revert "GetProviderConfig: generate YAML with string operations(!)"

This reverts commit 10cf9e9d70759396d32550b7a9d4aae700c43c51.

disq committed:
example as config

disq committed:
fix
